types/mmmagic/index.d.ts

```diff
@@ -5,13 +5,22 @@
 
 /// <reference types="node" />
 
+/**
+ * callback for detect() and detectFile()
+ * Result is a string, except when MAGIC_CONTINUE is set,
+ * then it is an array of string
+ */
+type DetectionCallback =
+    ((err: Error, result: string) => void) |
+    ((err: Error, result: string[]) => void) |
+    ((err: Error, result: string | string[]) => void)
```

Comment:
This is problematic, because it breaks inference for callback parameters that aren’t explicitly typed:

```ts
magic.detect(buf, (err, result) => {
                // ^^^  ^^^^^^ both implicitly 'any'
```

The strictest version is just to call it `(err: Error, result: string | string[]) => void`, and that would keep inference, but is a bit cumbersome and would be a real breaking change. There’s just no good way to type an API like this, unfortunately. I’m not sure what to suggest 😕

Reply:
You are right, I didn't think about untyped callbacks. One way to make this work would be to make either the `Magic` object, or its `detect` method generic, defaulting to the most common use-case which is `string` result. Something like this:

    class Magic {
      // ...
      magic<ResultType = string extends (string | string[])>(data: Buffer, callback: (err: Error, result: ResultType) => void)
      // ...
    }

This isn't ideal because nothing prevents you from using an incorrect `ResultType` (with respect to the flags passed to the constructor at runtime). But it's the only backward-compatible solution that I see.

If you think this idea is OK, I will amend my pull request and add tests for untyped callbacks.

Reply:
I talked to @sandersn about this, and we’re leaning toward just making the existing signature say `string | string[]`. It would be a “breaking change,” but this library is not super widely used, and the break is easy to fix. Ultimately it’s the only solution that’s _safe_, even though it’s _too_ safe.